File: python/day15.py
from pprint import pprint
import os
import time


dirs = { 
    '^' : (-1,0),
    'v' : (1,0),
    '<' : (0,-1),
    '>' : (0,1)
}


def print_warehouse(w, bot):
    print("\033[H\033[J", end="")
    br,bc = bot
    w[br][bc] = '@'
    for l in w:
        print("".join(l))
    w[br][bc] = '.'    

def move_box(warehouse, box, move):
    br,bc = box    
    dr,dc = move
    nr,nc = br+dr,bc+dc
    if warehouse[nr][nc] == '.':
        warehouse[nr][nc] = 'O'
        warehouse[br][bc] = '.'
        return True
    if warehouse[nr][nc] == '#':
        return False
    if warehouse[nr][nc] == 'O' and move_box(warehouse, (nr, nc), move):
        warehouse[nr][nc] = 'O'
        warehouse[br][bc] = '.'
        return True
    return False

def run(warehouse, moves, bot):
    br,bc = bot
    for c,m in enumerate(moves):
        #print_warehouse(warehouse, (br,bc))
        #print(f"{c} {m}")
        #time.sleep(0.01)
        dr,dc = m
        nr,nc = br+dr,bc+dc
        if warehouse[nr][nc] == '.':
            br,bc = nr,nc
            continue
        if warehouse[nr][nc] == '#':
            continue
        if move_box(warehouse, (nr,nc), m):
            br,bc = nr,nc

def gps(warehouse):
    acc = 0
    for r in range(len(warehouse)):
        for c in range(len(warehouse[r])):
            if warehouse[r][c] == 'O':
                acc += (100*r) + c
    return acc


def day15(input):
    warehouse = []
    for r,l in enumerate(input):
        if len(l) == 0:
            break
        line = []
        for c,m in enumerate(l):
            if m == '@':
                line.append('.')
                bot = (r,c)
            else:
                line.append(m)
        warehouse.append(line)
    
    moves = []
    for l in input[r:]:
        moves += [dirs[c] for c in l]
    run(warehouse, moves, bot)
    print(gps(warehouse))


input = open("src/bin/inputs/day15.test1.txt", "r").read().splitlines()
pprint(day15(input))

input = open("src/bin/inputs/day15.txt", "r").read().splitlines()
pprint(day15(input))

File: src/bin/day15.rs
const INPUT: &str = include_str!("inputs/day15.txt");

fn day15(input: &str) -> (usize, usize) {
    (input.len(), input.len())
}

fn main() {
    let (p1, p2) = day15(INPUT);
    println!("{}\n{}", p1, p2);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test1() {
        let input = include_str!("inputs/day15.test1.txt");
        let (p1, p2) = day15(input);

        assert_eq!(p1, input.len());
        assert_eq!(p2, input.len());
    }

    #[test]
    fn test_main() {
        let (p1, p2) = day15(INPUT);

        assert_eq!(p1, INPUT.len());
        assert_eq!(p2, INPUT.len());
    }
}
